### DISCUSSION TITLE
XPT2046 with SSD1963 (parallel)

### DISCUSSION THREAD

**Discussion by sn00zerman** [Bodmer/TFT_eSPI#1033] [Compatible support libraries]
Hi all,

I switched to a new display for my locomotion project. ( https://www.digitalplayground.be/locomotion.html )
We started this project with an ILI9488 (3.5" - 320x480), but we needed extra display-space, so we switched to SSD1963 (5" - 480x800)
in combination with ESP32, this works perfect. However, we are getting a bit short on pins :-)
We have come a long way with this project. (It's actually our first Arduino project, and if you look a the result at out website, it's a pity that we are blocked with one issue left ...)

- Almost all our extra peripheral runs on I2C (in IO21 & IO22 of ESP32 - push-buttons, joystick module (ADC) and rotary encoder)
- We need the serial port to send screenshots to the PC (IO1 & IO3)
- Almost all pins are occupied by the 8-bit parallel connection of the SSD1963
- We have an SD-card attached that also works fine

- We are currently struggling with the touch-screen. (also an XPT2046, like on the ILI9488)
- However, we use alternative pins for SPI,
```
  SPIClass * hspi = new SPIClass(HSPI);
  hspi->begin(23, 34, 32, 5);   //SCLK, MISO, MOSI, CS  
  SD.begin(5, *hspi);      

```
We still have GPIO 0 available for the CS of the touch-part of the screen, 
and would like to use the same SPI pins for the rest of the touch-interface.
We already did have a look at a lot of XPT2406 libraries (Utouch, Urtouch, ThingPulse, and the library from Paul Stoffregen)

Can someone help us out with a modified Extensions/Touch.cpp or an indication of using one of the afore mentioned libs and how to make them work with our SPI defined for our SD card ?

Thanks in advance,
Kris

**Comment by Bodmer** [comment]
I read the above twice but still do not understand how many or what questions are being asked! Here is a guess...

If you are short of pins, you may be able to free up pins by trying the following:
1. Do not connect the display RD pin to ESP32, instead tie the line to +3.3V
2. Connect the display RST pin the the ESP32 RST pin, or you may be able to just onnect to +3.3V as a software reset command is sent to the display during the tft.begin/init().
3. Do not connect the display CS line to the ESP32, instead connect to 0V so display is permanently enabled

If the question is about a touch library then there are several independant ones available, for example [this one](https://github.com/PaulStoffregen/XPT2046_Touchscreen) I am sure will work if correctly invoked.

**Reply by Bodmer** [reply]
[This one](https://github.com/achillhasler/TFT_eTouch) looks promising too.

**Comment by Bodmer** [comment]
Wrong link above, [this is the one](https://github.com/Bodmer/TFT_Touch) I meant to link to as the pins for DIN.DOUT etc are user defined. Not sure if it will work with the SD library though as the pins are bit bashed.

**Comment by Bodmer** [comment]
I had a really old copy of TFT_eTouch and see it has progressed to use effective filtering. I updated my copy and see it works very well with no false signals. I think that will be my "goto" touch library for the cheaper resistive touch screens.  The capacitive touch controller give a much nicer user touch experience (the type used on mobile phones) should you take the project to another stage.

The TFT_eTouch  library is probably the best solution for you at the moment as you can pass the constructor a pointer to the SPI class, so will be straightforward to incorporate in your existing code. E.g this should work:
```
SPIClass * hspi = new SPIClass(HSPI);
TFT_eTouch<TFT_eSPI> touch(tft, TFT_ETOUCH_CS, 0xff, *hspi);

...
void setup()
...
  hspi->begin(23, 34, 32, 5);   //SCLK, MISO, MOSI, CS  

```

Post back if this works as other parallel screen users may be interested.

**Reply by sn00zerman** [reply]
Hi Bodmer,


Thanks for your feedback. the RD pin did the trick, I now have enough free pins for the rest of my peripherals :-)

I did some testing, I have the SD and Touchscreen on the same SPI bus, SD works fine.

- With the PaulStoffRegen, i got the most "promising" results.
  I got back when the touch was touched or not, but the X and Y results, are "off".
  The X was only changing in the range of less then 100 !
  While moving the Y position, the X also changes range (changing more then 100 !)
- With the E_touch library from achillhasler, I get 0 for every value (X,Y,Z1,Z2,RZ)
- I tried the Bodmer library, I have no issue with the definition of the Pins, they do not conflict with the real
  SD communication, but they also result in giving back all zeroes ...

When I "move" everything to a smaller screen (3.5"), everything works fine.
It seems that the XPT2046 "reacts" different on this 5" screen ... 
(I don't have time yet to read the data-sheet, but I assume that there must be something "different")

**Comment by sn00zerman** [comment]
I tried several libraries until they returned a value.
They all give me the same problem: 
- display "touch" is returned, while nothing is touched
- only x OR y axis gives back a "wide range" of minimum-maximum data, not both axis at the same time.
- the first centimeter of the touch-panel, gives "0".

While I was actually short of one pin more that I needed, I decided the go "another" way with the touch-panel.
I ordered an STMPE610, this is an IC where you can connect the X+,X-,Y+,Y- of the raw display to, and this returns data over I2C.
I expect these controllers to arrive be next weekend, I keep you all informed.

best regards,
Kris

**Reply by Bodmer** [reply]
Odd that the touch does not work for so many different approaches.

Post a link to the vendors page of the display you bought and the hardware options chosen (if any) and I will look to see if I can spot any GOTCHAS!

**Comment by Bodmer** [comment]
I think the "calibrate" sketch of [this library](https://github.com/achillhasler/TFT_eTouch) should work but the library assumes pin 3 for the touch CS, so if you are still using pin 5 or another pin then you need to tweak the sketch or the[ library setup file](https://github.com/achillhasler/TFT_eTouch/blob/master/TFT_eTouchUser.h#L24).

**Reply by Bodmer** [reply]
The TFT_eTouch library has been tested with parallel displays it seems. There is a[ specific line here](https://github.com/achillhasler/TFT_eTouch/blob/master/examples/generic/calibrate/calibrate.ino#L34).

**Reply by Bodmer** [reply]
[This sketch](https://github.com/achillhasler/TFT_eTouch/blob/master/examples/generic/raw/raw.ino) can be configured to run without a display. So this would only use the SPI bus, surely that will run?
